components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/WebAppAuthenticatorImpl.java

```diff
@@ -56,10 +59,16 @@ public boolean authenticate(Message message) throws APIManagementException {
                 PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext.getThreadLocalCarbonContext();
                 RealmService realmService = (RealmService) carbonContext.getOSGiService(RealmService.class, null);
                 try {
+                    String username = tokenInfo.getEndUserName();
+                    if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                        if (username.endsWith(SUPER_TENANT_SUFFIX)) {
+                            username = username.substring(0, username.length() - SUPER_TENANT_SUFFIX.length());
+                        }
+                    }
                     tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
                     carbonContext.setTenantDomain(tenantDomain);
```

Comment:
Yes Krishanthan. The problem we had with starting a tenant flow is to decide where we should end it. Because this method; authenticate() is used in CXF Interceptor level, we cannot use the usual try/finally blocks and end it in a finally block.